File: android/src/main/kotlin/io/carrotquest/carrotquest_sdk/CarrotquestSdkPlugin.kt
package io.carrotquest.carrotquest_sdk

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.util.Log
import androidx.annotation.NonNull
import com.google.firebase.messaging.RemoteMessage
import io.carrotquest_sdk.android.Carrot
import io.carrotquest_sdk.android.Carrot.Callback
import io.carrotquest_sdk.android.core.main.ThemeSdk
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import org.json.JSONObject

/** CarrotquestSdkPlugin */
class CarrotquestSdkPlugin: FlutterPlugin, MethodCallHandler, ActivityAware {
    private lateinit var channel: MethodChannel
    private var context: Context? = null
    private var activity: Activity? = null

    private var pluginInitted = false

    private var appId: String? = null
    private var apiKey: String? = null
    private var isLightTheme: Boolean? = null


    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "carrotquest_sdk")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        if (call.method == "setup") {
            if (pluginInitted) {
                result.error("Plugin is already initialized.", null, null)
                return
            }

            setup(call, result)
            return
        }

        if(call.method == "auth") {
            auth(call, result)
            return
        }

        if(call.method == "logOut") {
            logOut(call, result)
            return
        }

        if (call.method == "sendToken") {
            sendToken(call, result)
            return
        }

        if (call.method == "sendFirebasePushNotification") {
            sendFirebasePushNotification(call, result)
            return
        }

        if (call.method == "openChat") {
            openChat(call, result)
            return
        }

        if(call.method == "setUserProperty") {
            setUserProperty(call, result)
            return
        }

        if(call.method == "trackEvent") {
            trackEvent(call, result)
            return
        }

        if(call.method == "getUnreadConversationsCount") {
            getUnreadConversationsCount(call, result)
            return
        }

        if (call.method == "getPlatformVersion") {
            result.success("Android ${android.os.Build.VERSION.RELEASE}")
            return
        }

        if(call.method == "pushNotificationsUnsubscribe") {
            pushNotificationsUnsubscribe(result)
            return
        }

        if(call.method == "pushCampaignsUnsubscribe") {
            pushCampaignsUnsubscribe(result)
            return
        }

        result.notImplemented()
    }

    // override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    //   channel.setMethodCallHandler(null)
    // }

    private fun checkPluginInitiated(@NonNull result: MethodChannel.Result): Boolean {
        if (!pluginInitted) {
            result.error(
                "The plugin hasn't been initialized yet. Do Carrot.io.carrotquest.carrotquest_sdk.setup(...) first .",
                null,
                null
            )
            return false
        }
        return true
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        pluginInitted = false
        context = null
    }

    override fun onDetachedFromActivity() {
        pluginInitted = false
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
    }

    //=====
    private fun setup(
        @NonNull call: MethodCall,
        @NonNull result: MethodChannel.Result,
    ) {

        apiKey = call.argument<String?>("api_key")
        appId = call.argument<String?>("app_id")
        isLightTheme = call.argument<Boolean?>("is_light_theme")
        if (apiKey == null || appId == null) {
            result.error("An error has occurred, the apiKey or appId is null.", null, null)
            return
        }

        val con = context
        if (con != null) {
            Carrot.setup(con, apiKey!!, appId!!, object : Callback<Boolean> {
                override fun onResponse(resultSetup: Boolean?) {
                    pluginInitted = (resultSetup == true)
                   
                    try {
                        if(resultSetup == true) {
                            Carrot.setNotificationIcon(R.drawable.ic_cqsdk_notification)
                            Carrot.setUnreadConversationsCallback(object : Callback<List<String>>{
                                override fun onResponse(unreadConversationsIds: List<String>?) {
                                    channel.invokeMethod("unreadConversationsCount", unreadConversationsIds?.size ?: 0)
                                }

                                override fun onFailure(t: Throwable?) {
                                    
                                }
                            })
                            if (isLightTheme == true) {
                                Carrot.setTheme(ThemeSdk.DARK)
                            } else {
                                Carrot.setTheme(ThemeSdk.LIGHT)
                            }
                            result.success(null)
                        } else {
                            result.error("Setup is failed", null, null)
                        }
                    } catch (e: java.lang.Exception) {
                        //result.error("Setup is failed", null, null)
                    }
                }

                override fun onFailure(t: Throwable?) {
                    pluginInitted = false
                    result.error("Setup is failed: " + t.toString(), null, null)
                }
            })
        } else {
            result.error("Context is null", null, null)
        }
    }

    private fun auth(@NonNull call: MethodCall,
        @NonNull result: MethodChannel.Result) {
        if (!checkPluginInitiated(result)) {
            return
        }

        val userId = call.argument<String?>("user_id")
        val userAuthKey = call.argument<String?>("user_auth_key")
        if (userId == null || userAuthKey == null) {
            result.error("An error has occurred, the userId or userAuthKey is null.", null, null)
            return
        }

        Carrot.auth(userId, userAuthKey, object : Callback<Boolean> {
            override fun onResponse(resultAuth: Boolean?) {
                if(resultAuth == true) {
                    result.success(null)
                } else {
                    result.error("Auth is failed", null, null)
                }
            }

            override fun onFailure(t: Throwable?) {
                result.error("Auth is failed: " + t.toString(), null, null)
            }
        })
    }

    private fun logOut(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        if (!checkPluginInitiated(result)) {
            return
        }

        if (apiKey == null || appId == null) {
            result.error("An error has occurred, the apiKey or appId is null.", null, null)
            return
        }

        Carrot.deInit(object : Callback<Boolean>{
            override fun onResponse(resDeInit: Boolean) {
                if(!resDeInit) {
                    result.error("deInit is failed", null, null)
                    return
                }

                val con = context
                if (con != null) {
                    Carrot.setup(con, apiKey!!, appId!!, object : Callback<Boolean> {
                        override fun onResponse(resultSetup: Boolean?) {
                            pluginInitted = (resultSetup == true)

                            try {
                                if(resultSetup == true) {
                                    result.success(null)
                                } else {
                                    result.error("Setup is failed", null, null)
                                }
                            } catch (e: java.lang.Exception) {
                                //result.error("Setup is failed", null, null)
                            }
                        }

                        override fun onFailure(t: Throwable?) {
                            pluginInitted = false
                            result.error("Setup is failed: " + t.toString(), null, null)
                        }
                    })
                } else {
                    result.error("Context is null", null, null)
                }
            }

            override fun onFailure(t: Throwable) {
                result.error(t.localizedMessage, null, null)
            }
        })
    }

    private fun sendToken(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        if (!checkPluginInitiated(result)) {
            return
        }

        val token = call.argument<String?>("token")
        Carrot.sendFcmToken(token)

        result.success(null)
    }

    private fun sendFirebasePushNotification(
        @NonNull call: MethodCall,
        @NonNull result: MethodChannel.Result
    ) {
        val bundle = Bundle()
        val data: Map<String, Any> =
            call.argument<Map<String, Any>>("data") ?: HashMap<String, Any>()
            
        //Bundle().put
        val message = RemoteMessage(bundle)
        bundle.putString("id", data["id"]?.toString() ?: "")
        bundle.putString("title", data["title"]?.toString() ?: "")
        bundle.putString("body", data["body"]?.toString() ?: "")
        bundle.putString("conversation_id", data["conversation_id"]?.toString() ?: "")
        bundle.putString("conversation_type", data["conversation_type"]?.toString() ?: "")
        bundle.putString("sent_via", data["sent_via"]?.toString() ?: "")
        bundle.putString("body_json", data["body_json"]?.toString() ?: "")
//        bundle.putBoolean(
//            "is_carrot",
//            java.lang.Boolean.parseBoolean(data["is_carrot"]?.toString() ?: "False")
//        )
        bundle.putString("is_carrot", data["is_carrot"]?.toString() ?: "")
        bundle.putString("direction", data["direction"]?.toString() ?: "")

        Carrot.sendFirebasePushNotification(message, if(activity == null) context else activity)

        result.success(null)
    }


    private fun openChat(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        if (!checkPluginInitiated(result)) {
            return
        }
        try {
            if (activity != null) {
                Carrot.openChat(activity)
                result.success(null)
            } else {
                result.error("Activity in null", null, null)
            }
        } catch (e: Exception) {
            result.error(e.localizedMessage, null, null)
        }
    }

    private fun setUserProperty(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        if (!checkPluginInitiated(result)) {
            return
        }
        try {
            val key = call.argument<String?>("key")
            val value = call.argument<String?>("value")
            if(key == null || value == null) {
                result.error("Key or value is empty", null, null)
                return
            }

            Carrot.setUserProperty(key, value.toString())
            result.success(null)
        } catch (e: Exception) {
            result.error(e.localizedMessage, null, null)
        }
    }

    private fun trackEvent(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        if (!checkPluginInitiated(result)) {
            return
        }
        try {
            val event = call.argument<String?>("event")
            if(event == null) {
                result.error("Event is empty", null, null)
                return
            }

            val paramsStr = call.argument<String?>("params")
            if(paramsStr == null || paramsStr.isEmpty()) {
                Carrot.trackEvent(event)
                result.success(null)
            } else {
                Carrot.trackEvent(event, paramsStr)
                result.success(null)
            }

        } catch (e: Exception) {
            result.error(e.localizedMessage, null, null)
        }
    }

    private fun getUnreadConversationsCount(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        if (!checkPluginInitiated(result)) {
            return
        }
        try {
            val count =  Carrot.getUnreadConversations().size
            result.success(count)
        } catch (e: Exception) {
            result.error(e.localizedMessage, null, null)
        }
    }

    private fun pushNotificationsUnsubscribe(@NonNull result: MethodChannel.Result) {
        try {
            Carrot.pushNotificationsUnsubscribe()
            result.success(null)
        } catch (e: Exception) {
            result.error(e.localizedMessage, null, null)
        }
    }

    private fun pushCampaignsUnsubscribe(@NonNull result: MethodChannel.Result) {
        try {
            Carrot.pushCampaignsUnsubscribe()
            result.success(null)
        } catch (e: Exception) {
            result.error(e.localizedMessage, null, null)
        }
    }
}

File: ios/Classes/CarrotquestSdkPlugin.swift
import Flutter
import UIKit
import CarrotSDK
import Photos

public class CarrotquestSdkPlugin: NSObject, FlutterPlugin {
  private var channel: FlutterMethodChannel? = nil

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "carrotquest_sdk", binaryMessenger: registrar.messenger())
    let instance = CarrotquestSdkPlugin()
    instance.channel = channel
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
      switch call.method {
      case "getPlatformVersion":
          result("iOS " + UIDevice.current.systemVersion)
      case "setup":
          self._setup(with: call, and: result)
      case "auth":
          self._auth(with: call, and: result)
      case "logOut":
          self._logOut(with: call, and: result)
      case "sendToken":
          self._sendPushToken(with: call, and: result)
      case "sendFirebasePushNotification":
          self._sendFirebasePushNotification(with: call, and: result)
      case "setUserProperty":
          self._setUserProperty(with: call, and: result)
      case "trackEvent":
          self._trackEvent(with: call, and: result)
      case "getUnreadConversationsCount":
          self._getUnreadConversationsCount(with: result)
      case "openChat":
          self._openChat(with: result)
      case "pushNotificationsUnsubscribe":
          self._pushNotificationsUnsubscribe()
      case "pushCampaignsUnsubscribe":
          self._pushCampaignsUnsubscribe()
      default:
          result(FlutterMethodNotImplemented)
      }
  }

  private func _setup(with call: FlutterMethodCall, and result: @escaping FlutterResult) {
    guard let args = call.arguments as? NSDictionary else { return }
    guard let apiKey = args["api_key"] as? String else { return }
    let appGroup = args["app_group"] as? String
    guard let isLightMode = args["is_light_mode"] as? Bool else { return }
    Carrot.shared.setup(
        withApiKey: apiKey,
        withTheme: isLightMode ? Carrot.Theme.light : Carrot.Theme.dark,
        withAppGroup: appGroup,
        successHandler: {
            Carrot.shared.getUnreadConversationsCount { [weak self] count in
                guard let self = self else { return }
                self.channel?.invokeMethod("unreadConversationsCount", arguments: count)
            }
            result(nil)
        },
        errorHandler: { e in
            result(FlutterError(code: "setup", message: "setup error", details: e))
        }
    )
  }

    private func _openChat(with result: @escaping FlutterResult) {
      Carrot.shared.openChat()
      result(nil)
    }

    private func _auth(with call: FlutterMethodCall, and result: @escaping FlutterResult) {
      guard let args = call.arguments as? NSDictionary else { return }
      guard let userAuthKey = args["user_auth_key"] as? String else { return }
        guard let userId = args["user_id"] as? String else {return}

        Carrot.shared.auth(withUserId: userId, withUserAuthKey: userAuthKey, successHandler: {
            result(nil)
        }, errorHandler: { e in result(FlutterError(code: "auth", message: "auth error", details: e))})
    }

    private func _setUserProperty(with call: FlutterMethodCall, and result: @escaping FlutterResult) {
      guard let args = call.arguments as? NSDictionary else { return }
      guard let keyProperty = args["key"] as? String else { return }
      guard let valueProperty = args["value"] as? String else {return}
      let userProperties:[UserProperty] = [UserProperty(key: keyProperty, value: valueProperty)]

        Carrot.shared.setUserProperty(userProperties)
        result(nil)
    }

    private func _trackEvent(with call: FlutterMethodCall, and result: @escaping FlutterResult) {
      guard let args = call.arguments as? NSDictionary else { return }
      guard let event = args["event"] as? String else { return }
      guard let params = args["params"] as? String else {
          Carrot.shared.trackEvent(withName: event, withParams: "")
          result(nil)
          return
      }

        Carrot.shared.trackEvent(withName: event, withParams: params)
        result(nil)
    }

    private func _getUnreadConversationsCount(with result: @escaping FlutterResult) {
        Carrot.shared.getUnreadConversationsCount({count in result(count)})
    }

    private func _logOut(with call: FlutterMethodCall, and result: @escaping FlutterResult) {
        Carrot.shared.logout(successHandler: {
            result(nil)
        }, errorHandler: { e in result(FlutterError(code: "logOut", message: "logOut error", details: e))})
    }

    private func _sendPushToken(with call: FlutterMethodCall, and result: @escaping FlutterResult) {
        guard let args = call.arguments as? NSDictionary else { return }
        guard let token = args["token"] as? String else { return }
        CarrotNotificationService.shared.setToken(token)
        result(nil)
    }

    private func _sendFirebasePushNotification(with call: FlutterMethodCall, and result: @escaping FlutterResult) {
        guard let args = call.arguments as? NSDictionary else { return }
        guard let userInfo = args["data"] as? [String: Any] else { return }
//          CarrotNotificationService.shared.show(from: userInfo)
        result(nil)
    }

    private func _pushNotificationsUnsubscribe() {
        CarrotNotificationService.shared.pushNotificationsUnsubscribe()
    }

    private func _pushCampaignsUnsubscribe() {
        CarrotNotificationService.shared.pushCampaignsUnsubscribe()
    }
}

